librz/type/path.c

```diff
@@ -36,74 +36,167 @@ RZ_API void rz_type_path_free(RZ_NULLABLE RzTypePath *tpath) {
 	free(tpath);
 }
 
+static RzType *path_walker_parse_bracket(const RzTypeDB *typedb, RzType *parent, const char *path, size_t *i, st64 *offset) {
+	size_t nd = 0;
+	RzType *typd = parent;
+	st64 cur_dim_off = 1; // product of all dimension sizes
+	while (typd->kind == RZ_TYPE_KIND_ARRAY) {
+		++nd;
+
+		if (typd->array.count != 0 && cur_dim_off <= ST64_MAX / typd->array.count) { // check for overflow
+			cur_dim_off *= typd->array.count;
+		} else {
+			RZ_LOG_ERROR("Too many error subscriptions\n");
+			*offset = -1;
+			return NULL;
+		}
+
+		typd = typd->array.type;
+	}
+	cur_dim_off *= rz_type_get_base_type(typedb, typd)->size; // elem size in bits
+
+	typd = parent;
+	for (size_t id = 0; id < nd; ++id) {
+		if (path[*i] != '[') {
+			RZ_LOG_ERROR("Expected '[' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		++*i;
+		size_t tok_beg = *i;
+		for (; isdigit(path[*i]); ++*i)
+			;
+
+		if (path[*i] != ']') {
+			RZ_LOG_ERROR("Expected ']' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		char *idx_str = rz_str_ndup(&path[tok_beg], *i - tok_beg);
+		size_t idx = rz_num_math(NULL, idx_str);
+		free(idx_str);
+
+		cur_dim_off /= typd->array.count;
+		*offset += cur_dim_off * idx;
+		typd = typd->array.type;
+
+		// DONT check for "idx < array size"
+		// because guy may be exploiting OOB or something else.
+		// so just continue.
+
+		++*i; // skip ']'
+	}
+
+	return typd;
+}
+
+static RzType *path_walker_parse_dot(const RzTypeDB *typedb, RzType *parent, const char *path, size_t *i, st64 *offset) {
+	++*i;
+
+	size_t tok_beg = *i;
+	for (; isalnum(path[*i]); ++*i)
+		;
+
+	char *tok = rz_str_ndup(&path[tok_beg], *i - tok_beg);
+
+	RzBaseType *parent_btype = rz_type_get_base_type(typedb, parent);
+	if (!parent_btype) {
+		RZ_LOG_ERROR("Could not found btype for parent '%s'\n", parent->identifier.name);
+		*offset = -1;
+		free(tok);
+		return NULL;
+	}
+
+	RzTypeStructMember *memb_it;
+	RzType *cur_type = NULL;
+	size_t cur_offset = -1;
+	rz_vector_foreach(&parent_btype->struct_data.members, memb_it) {
+		if (!strcmp(memb_it->name, tok)) {
+			cur_type = memb_it->type;
+			cur_offset = memb_it->offset; // in bytes
+			cur_offset *= 8; // in bits
+			break;
+		}
+	}
+
+	if (!cur_type) {
+		RZ_LOG_ERROR("Invalid member '%s' for parent type\n", tok);
+		*offset = -1;
+		free(tok);
+		return NULL;
+	}
+
+	free(tok);
+
+	*offset += cur_offset;
+	if (path[*i] == '[') {
+		parent = cur_type;
+		if (parent->kind != RZ_TYPE_KIND_ARRAY) {
+			RZ_LOG_ERROR("Expected array, got another type\n");
+			*offset = -1;
+			return NULL;
+		}
+
+		path_walker_parse_bracket(typedb, parent, path, i, offset);
+		if (*offset == -1) {
+			return NULL;
+		}
+
+		// unwrap array
+		while (parent->kind == RZ_TYPE_KIND_ARRAY) {
```

Comment:
Sorry, haven't looked at the semantic. But is this guaranteed to terminate? Meaning, will in the `parent->array.type` sequence always be a node with `RZ_TYPE_KIND_ARRAY`? Or are malfunctions possible?

Reply:
It seems I don't really understand what you've meant. The loop breaks if current node is not `RZ_TYPE_KIND_ARRAY`.

Reply:
Ah, ignore it. I was obviously diluted :D